x/concentrated-liquidity/swaps_tick_cross_test.go

```diff
@@ -878,6 +880,199 @@ func (s *SwapTickCrossTestSuite) TestSwapOutGivenIn_Tick_Initialization_And_Cros
+		// This flag is used to control an edge case behavior in test setup
+		// when swapping right and then right again within the same tick.
+		// It is only used to signal the swap direction while estimating the expected results.
+		isOneForZeroWithinSameTick bool
+
+		// isPositionActiveFlag is used to control the expected state of position
+		// at the end of configured swaps of the test.
+		// See diagram above the definition of defaultTickSpacingsAway variable for layout.
+		// The first position is NR1, the second position is NR2 etc.
+		// That is, the wider range position is preceeds the narrower range position.
+		isPositionActiveFlag []bool
+	}
+
+	// computeExpectedTicks estimates the expected tick values from the given test case
+	// and returns them as tick slice.
+	computeExpectedTicks := func(poolId uint64, tc continugousTestCase) []int64 {
+		pool, err := s.App.ConcentratedLiquidityKeeper.GetPoolById(s.Ctx, poolId)
+		s.Require().NoError(err)
+
+		originalCurrentTick := pool.GetCurrentTick()
+
+		// Convert the given "ticks away from original current" to actual tick indexes.
+
+		expectedSwapEndTicks := make([]int64, len(tc.swapEndTicksAwayFromOriginalCurrent))
+
+		for i, swapTicksAway := range tc.swapEndTicksAwayFromOriginalCurrent {
+			// We special case the min and max tick to be given by absolute values
+			// rather than relative to the original current tick.
+			if swapTicksAway == types.MinInitializedTick {
+				// In zero for one direction, we kick the tick back by one while crossing.
+				// This is to ensure that our definition of "active bucket" is correct.
+				expectedSwapEndTicks[i] = types.MinCurrentTick
+			} else if swapTicksAway == types.MaxTick {
+				expectedSwapEndTicks[i] = types.MaxTick
+			} else {
+				expectedSwapEndTicks[i] = originalCurrentTick + swapTicksAway
+			}
+		}
+
+		return expectedSwapEndTicks
+	}
+
+	// computeExpectedValuesForTestOneForZero returns the tick to swap to during estimate computation and amountIn multiplier.
+	// It most cases, the tick to swap to is the same as the expected tick to reach after the swap and the multiplier is 1.
+	// The only exception is when performing a second swap in the same direction within the same tick.
+	// In such a case, we need to run our estimate logic one tick further to ensure that our estimate is non-zero.
+	// However, we discount the amountIn by half to ensure that the tick is not crossed.
+	computeNextTickToReachAndMultiplier := func(isZeroForOne bool, expectedSwapEndTick int64, shouldStayWithinTheSameTickInCompute bool) (int64, sdk.Dec) {
+		if shouldStayWithinTheSameTickInCompute {
+			nextTickToReachInCompute := expectedSwapEndTick
+			if isZeroForOne {
+				nextTickToReachInCompute = nextTickToReachInCompute - 1
+			} else {
+				nextTickToReachInCompute = nextTickToReachInCompute + 1
+			}
+
+			return nextTickToReachInCompute, sdk.NewDecWithPrec(5, 1)
+		}
+
+		return expectedSwapEndTick, sdk.OneDec()
+	}
```

Comment:
Can you explain a little more about the multiplier here? I later realized we used 50 percent to stay within the same tick, but it was not clear when initially reading the code here.

Reply:
You are correct. We use a 50% discount to stay within the same bucket. 50% picked arbitrarily. If we don't discount, the next tick will be crossed as we are running the token in estimation logic between ticks. For some test cases, we specifically want to test staying within the same bucket.

Let me know if there is anything else that I should elaborate on

Reply:
Can you add that the 50 percent is arbitrary in the comments to make it easier for other reviewers?

Reply:
done